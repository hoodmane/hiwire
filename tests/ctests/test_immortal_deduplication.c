#include "testlib.h"

int
main()
{
  __externref_t o1 = get_obj(7);
  HwRef id1 = hiwire_new(o1);
  ASSERT(!SLOT_DEDUPLICATED(id1));
  HwRef id2 = hiwire_incref_deduplicate(id1);
  ASSERT(SLOT_DEDUPLICATED(id1));
  ASSERT(id1 == id2);
  HwRef id_immortal = hiwire_intern(o1);
  HwRef id3 = hiwire_incref_deduplicate(id_immortal);
  ASSERT(id3 == id1);
  ASSERT(SLOT_REFCOUNT(id1) == 3);

  hiwire_decref(id1);
  hiwire_decref(id1);
  hiwire_decref(id1);

  HwRef id4 = hiwire_incref_deduplicate(id_immortal);
  ASSERT(id4 == id_immortal);

  HwRef id5 = hiwire_new(o1);
  HwRef id6 = hiwire_incref_deduplicate(id5);
  ASSERT(!SLOT_DEDUPLICATED(id5));
  ASSERT(id6 == id_immortal);
  ASSERT(SLOT_REFCOUNT(id5) == 1);

  return 0;
}
