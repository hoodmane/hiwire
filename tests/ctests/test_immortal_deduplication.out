[94;1mPASS [0m   !SLOT_DEDUPLICATED(id1)
[94;1mPASS [0m   SLOT_DEDUPLICATED(id1)
[94;1mPASS [0m   id1 == id2
[94;1mPASS [0m   id3 == id1
[94;1mPASS [0m   SLOT_REFCOUNT(id1) == 3
[94;1mPASS [0m   id4 == id_immortal
[94;1mPASS [0m   !SLOT_DEDUPLICATED(id5)
[94;1mPASS [0m   id6 == id_immortal
[94;1mPASS [0m   SLOT_REFCOUNT(id5) == 1
